compiler/rustc_error_messages/src/lib.rs

```diff
@@ -539,3 +541,47 @@ impl From<Vec<Span>> for MultiSpan {
         MultiSpan::from_spans(spans)
     }
 }
+
+#[macro_export]
+macro_rules! pluralize {
+    ($x:expr) => {
+        if $x != 1 { "s" } else { "" }
+    };
+    ("has", $x:expr) => {
+        if $x == 1 { "has" } else { "have" }
+    };
+    ("is", $x:expr) => {
+        if $x == 1 { "is" } else { "are" }
+    };
+    ("was", $x:expr) => {
+        if $x == 1 { "was" } else { "were" }
+    };
+    ("this", $x:expr) => {
+        if $x == 1 { "this" } else { "these" }
+    };
+}
+
+pub fn build_list_separated_with_and<T: writeable::Writeable>(
+    values: impl Iterator<Item = T> + Clone,
+    value_count: usize,
+    descr: &str,
+    spread_limit: usize,
+) -> String {
+    assert!(value_count > 0);
+
+    if value_count > spread_limit {
+        format!("multiple {descr}{s}", s = pluralize!(value_count),)
+    } else {
+        let data_provider = rustc_baked_icu_data::baked_data_provider();
+        let locale = icu_locid::locale!("en").into();
+        let list_formatter = icu_list::ListFormatter::try_new_and_with_length_unstable(
```

Comment:
thought: ideally the list formatter is built once and stored in the session or something

Reply:
basically it needs to be stored alongside the fluent stuff